### PR TITLE
 Add pwr.sum with sha256 digests

### LINKED ISSUE #9: Add checksum for tag sha256

### PR DESCRIPTION
**Pull request by tbateson**
Closes #9

### CODE REVIEW

**Comment by troyaf** [comment]
image


I think there might be an issue.

**Comment by tbateson** [comment]
> I think there might be an issue.

Looks like you pressed the enter key a lot? What did you do to cause this? I'm not able to reproduce.

**Comment by tbateson** [comment]
image

**Comment by troyaf** [comment]
When there is a `pwr.pkg+pwr.json` in the ancestry of calling a `pwr sh <pkg>`, we crash even if the `pwr.json` is not being used. I don't like that. I would prefer we only do the pkg check specifically when the config was loaded.

**Comment by troyaf** [comment]
Another case I think we should account for is when there is a PKG file committed and a use hasn't fetched the latest tags yet. We should make sure the version of a package that is loaded is the one in the sum file. If we allow newer versions without complaint but error on old versions, I think we'll get the behavior we want.

We should also account for overrides, and not verify overridden packages.

**Comment by tbateson** [comment]
> We should make sure the version of a package that is loaded is the one in the sum file. If we allow newer versions without complaint but error on old versions, I think we'll get the behavior we want.

It would probably be better behavior if we forced tags to update when a newer version is encountered in the pkg file, but I'm not sure how easily that could be done.

**Comment by troyaf** [comment]
Here's an odd scenario that happened:

1. I have `ninja:1.11.0` in the sum file.
1. I changed the config to load `ninja:1.10`, and it crashes as expected.
1. I changed the config to load `ninja:1.10.2`, and the console returns with no output.

File: src/pwr.ps1 [hunk 1]
```diff
@@ -804,18 +897,20 @@ if (-not $Run) {
...
 function Get-PwrConfig {
-	$PwrCfgDir = $ExecutionContext.SessionState.Path.CurrentLocation
-	do {
-		$PwrCfg = "$PwrCfgDir\pwr.json"
-		if (Test-Path -Path $PwrCfg -PathType Leaf) {
-			try {
-				return $PwrCfg, (Get-Content $PwrCfg | ConvertFrom-Json)
-			} catch {
-				Write-PwrFatal "bad JSON parse of file $PwrCfg - $_"
+	if (-not $Packages) {
```
Comment (troyaf) [inline src/pwr.ps1:900]: I'm pretty sure this broke overrides

**Review by troyaf — CHANGES_REQUESTED** [review]

**Review by troyaf — CHANGES_REQUESTED** [review]
Here's an odd scenario that happened:

1. I have `ninja:1.11.0` in the sum file.
1. I changed the config to load `ninja:1.10`, and it crashes as expected.
1. I changed the config to load `ninja:1.10.2`, and the console returns with no output.

**Comment by tbateson** [comment]
> Here's an odd scenario that happened

This was caused by a `string`/`SemanticVersion` type mismatch, fixed so that won't ever happen again.

File: src/pwr.ps1 (hunk 1 above)
Comment (tbateson) [reply src/pwr.ps1:941]: Oops, forgot override was a switch

Comment (tbateson) [reply src/pwr.ps1]: Should be fixed now.

**Comment by troyaf** [comment]
I'm getting `pwr: package ninja:1.10.2 is older than version 1.11.0 in C:\Users\Troy\code\airpwr\pwr.sum` probably because
`if ($V -and $V.LaterThan($PkgVersion)) {` is not isolated to just `$v.laterthan`.

**Comment by tbateson** [comment]
> I'm getting `pwr: package ninja:1.10.2 is older than version 1.11.0 in C:\Users\Troy\code\airpwr\pwr.sum` probably because `if ($V -and $V.LaterThan($PkgVersion)) {` is not isolated to just `$v.laterthan`.

So I'm not sure if we're thinking of the same issue, but I've basically made it so that it will only complain about the later-than issue if the pwr.sum is in sync with the pwr.json. Therefore if you have ninja:1.10 in your config it will only check versions that are related to that. E.g. if you have ninja:1.11.1 in pwr.sum it won't crash if you change your pwr.json to ninja:1.10.

**Pull request closed without merging**